Sidebar: Redesign upsell nudge

sixhours opened the pull request:
#### Changes proposed in this Pull Request

* Some color schemes use the same background and border color for the upsell nudge and the current highlighted menu item, making the two run together.
* This PR moves the upsell nudge into the site card area, adds a small margin around it and rounds the corners. Also removes the left colored border.

**Before**

<img width="330" alt="Screen Shot 2020-12-15 at 3 32 35 PM" src="https://user-images.githubusercontent.com/2124984/102270830-77520500-3eec-11eb-859e-d3e8b227a890.png">

**After**

<img width="290" alt="Screen Shot 2021-01-05 at 9 35 40 AM" src="https://user-images.githubusercontent.com/2124984/103662243-b4cf0e80-4f3d-11eb-9000-5d55c83b3c80.png"> <img width="292" alt="Screen Shot 2021-01-05 at 9 35 51 AM" src="https://user-images.githubusercontent.com/2124984/103662246-b567a500-4f3d-11eb-843c-47a0193ffa41.png"> <img width="294" alt="Screen Shot 2021-01-05 at 9 36 07 AM" src="https://user-images.githubusercontent.com/2124984/103662247-b567a500-4f3d-11eb-845c-61703849eaed.png"> <img width="291" alt="Screen Shot 2021-01-05 at 9 36 22 AM" src="https://user-images.githubusercontent.com/2124984/103662249-b567a500-4f3d-11eb-9d64-5722b69af72b.png">


#### Testing instructions

* Switch to this PR
* Switch color schemes in `/me/account` on a site that has a notice in the sidebar (Claim your free domain, for example)
* Go to My Home
* Check out the upsell nudge(s) and the new design.
* Switch to other color schemes to check the visuals.

Fixes #48367

sixhours committed:
Whoops, just my commits, please.

rickybanister commented:
I saw the space correctly in most of the color schemes I tried, but for some reason there's no space when using the 'classic dark' scheme. Ah, I see how you implemented this change.. is there any reason not to just add the margin for all color schemes?

sixhours commented:
The extra margin looks like a bug to me when I add it to all color schemes, particularly the lighter ones:

<img width="295" alt="Screen Shot 2020-12-15 at 4 18 53 PM" src="https://user-images.githubusercontent.com/2124984/102274117-58a23d00-3ef1-11eb-85b5-51c92cb97403.png">

<img width="298" alt="Screen Shot 2020-12-15 at 4 19 03 PM" src="https://user-images.githubusercontent.com/2124984/102274118-58a23d00-3ef1-11eb-80b2-ec537cd37ad4.png">

<img width="274" alt="Screen Shot 2020-12-15 at 4 18 23 PM" src="https://user-images.githubusercontent.com/2124984/102274113-5809a680-3ef1-11eb-92cf-545808dd05e5.png">

<img width="282" alt="Screen Shot 2020-12-15 at 4 18 43 PM" src="https://user-images.githubusercontent.com/2124984/102274116-58a23d00-3ef1-11eb-9c00-205ee4f303df.png">

krymson24 commented:
Wow, this is super subtle... even with the arrows on the screenshots 😆 . Honestly, I don't see anything wrong with your screenshots. The selective styling for schemes might make things more difficult to manage in the future, and could be ripe for unexpected styling. Up to you though.

getdave commented:
cc @Automattic/ajax to whom this bug was reported in p7DVsv-a0W-p2#comment-33082.

mmtr commented:
I concur with what @krymson24 said. Having the extra pixel only on certain themes may complicate the maintenance as new themes are added or existing ones are changed.

Would it be possible to update the affected themes so they use different colors for the nudge border and the highlighted nav item?

frontdevde commented:
> The selective styling for schemes might make things more difficult to manage in the future, and could be ripe for unexpected styling.

I'd strongly second this. Selective styling aside though, my main concern would be that the current approach introduces changes to properties other than colors. Color schemes should only affect the colors, changes to sizing etc could introduce a whole host of new issues (especially if applied selectively).

Also please note that any change to color schemes affecting the masterbar or sidebar on the Calypso side of things now also needs to be implemented on the wp-admin side of things due to nav unification.

sixhours commented:
> Would it be possible to update the affected themes so they use different colors for the nudge border and the highlighted nav item?

I was wondering about this, too, but I think we're trying to match core, and that complicates which values go with which parts of the nav/overall color scheme.

We could add 1px bottom border to the nudge for all schemes, mapped to a color in that scheme that works better than the background of the nav.

Or remove the border from the nudge entirely and let the CTA do the heavy lifting as a callout.

What do you think @rickybanister @frontdevde ?

rickybanister commented:
I'd be ok with removing the border and adjusting the shape of the nudge. I'll leave the details of that up to @sfougnier, but I do like his round rect nudge design with a bigger margin all around it. 

A solution that works the same for all color schemes is definitely preferred.

ghost commented:
> A solution that works the same for all color schemes is definitely preferred.

Agreed with this. If we change this nudge padding or styling it should impact all color schemes.

> I'd be ok with removing the border and adjusting the shape of the nudge.

I think perhaps adding some 4px or 8px margins around the nudge and adding some rounded corners could help, but one fix would be to simply remove the left border so things blend less. Here's what I tried in [Figma](https://www.figma.com/file/9cy0DIowQkPAu0rodb9jsL/WP-Admin-for-Dotcom?node-id=1274%3A900).

<img width="2033" alt="Screen Shot 2020-12-17 at 10 19 06 AM" src="https://user-images.githubusercontent.com/67021691/102506660-59ea7b80-4051-11eb-8de6-bf76f8c2a88e.png">

rickybanister commented:
I think those are a lot stronger. It probably doesn't need the left border, but I guess we can look at it with various color schemes to see how it works. The same is true for this treatment—the lighter color schemes will make or break how the margin around the nudge feels.

sixhours commented:
I think if we're rounding the corners, the border will look weird and we should remove it. Let me see what a lighter color scheme will do without the border and adding the rounded corners!

sixhours commented:
4px margin | 8px margin
----------- | ------------
<img width="285" alt="Screen Shot 2020-12-18 at 10 32 41 AM" src="https://user-images.githubusercontent.com/2124984/102632194-bc5c7e00-411c-11eb-860b-1b943a6006aa.png"> | <img width="282" alt="Screen Shot 2020-12-18 at 10 13 04 AM" src="https://user-images.githubusercontent.com/2124984/102632203-c1213200-411c-11eb-9d1b-6e1ec40145dc.png">

And some examples in other color schemes:

<img width="285" alt="Screen Shot 2020-12-18 at 10 34 54 AM" src="https://user-images.githubusercontent.com/2124984/102632413-12c9bc80-411d-11eb-891a-3f455a976d8e.png"> <img width="283" alt="Screen Shot 2020-12-18 at 10 35 57 AM" src="https://user-images.githubusercontent.com/2124984/102632416-13625300-411d-11eb-9d6f-08b7c06245d6.png">
<img width="284" alt="Screen Shot 2020-12-18 at 10 36 33 AM" src="https://user-images.githubusercontent.com/2124984/102632421-13fae980-411d-11eb-951f-b482f8eb6f6d.png"> <img width="295" alt="Screen Shot 2020-12-18 at 10 36 47 AM" src="https://user-images.githubusercontent.com/2124984/102632423-14938000-411d-11eb-9042-c80d36ce5236.png">
<img width="289" alt="Screen Shot 2020-12-18 at 10 37 08 AM" src="https://user-images.githubusercontent.com/2124984/102632425-14938000-411d-11eb-8778-8cb90a0e59a6.png"> <img width="288" alt="Screen Shot 2020-12-18 at 10 37 38 AM" src="https://user-images.githubusercontent.com/2124984/102632427-152c1680-411d-11eb-9512-596db9b53c51.png"> <img width="292" alt="Screen Shot 2020-12-18 at 10 38 08 AM" src="https://user-images.githubusercontent.com/2124984/102632462-24ab5f80-411d-11eb-8f93-70e2f0c1e5fa.png">

kwight commented:
FWIW, I find the thicker padded option with no left border to be the easiest on the eyes. The _best_ solution to this design problem is to remove these upsells from the sidebar completely, but I'll show myself out... 👋

ghost commented:
> FWIW, I find the thicker padded option with no left border to be the easiest on the eyes. The best solution to this design problem is to remove these upsells from the sidebar completely, but I'll show myself out

Agreed! 😆 

One small detail: I didn't realize there was a horizontal rule hiding behind the full-width nudge. The extra white space created by the added padding feels a bit trapped by the divider line and makes it feel a bit hard to focus the eye. Any chance of removing the hr and removing the top padding on the nudge so it's flush with the bottom of the site card?

rickybanister commented:
Agreed with Saul, but that might require us to stick the nudge array inside the sitecard div. In the event there isn't a nudge it could look weird without that border (but we should check to be sure).

sixhours commented:
We can move the border to the top of the menu and remove it from the site card so it works both when we have a notice and when we don't (though I need to test the second type of notice with this arrangement, it should be fine):

<img width="300" alt="Screen Shot 2020-12-18 at 2 39 54 PM" src="https://user-images.githubusercontent.com/2124984/102656113-5fbf8a00-4141-11eb-9b4e-a7597c2eddf9.png">

krymson24 commented:
Looks good if we don't want to intentionally separate the CTA from the site name and site!

sixhours committed:
Update sidebar upsell nudge styles

sixhours commented:
I've updated this with the new design. I still need to test it with the other upsell/nudge types displayed at the same time.

sixhours commented:
The domain-related nudges look pretty weird here: 

<img width="292" alt="Screen Shot 2021-01-05 at 10 10 37 AM" src="https://user-images.githubusercontent.com/2124984/103662754-50f91580-4f3e-11eb-8edc-7cd6eae26869.png">

I'll see if I can bring them into alignment with the new design.

cpapazoglou commented:
👋 there! Please bare in mind that the new color schemes are a side effect of the nav-unification efforts! This leads to 2 problems:

1) If the changes in this PR affect how the sidebar is shown, they will also need to be ported in wp-admin. Except if they are nudge/notices specific which are not shown currently in wp-admin

Here for the sidebar
https://github.com/Automattic/jetpack/blob/38e9ae6cf9c4a4d96a074f431b7f52ce5ea79f93/modules/masterbar/admin-menu/admin-menu.css#L0-L1

Or here for color-schemes
https://github.com/Automattic/jetpack/blob/7de87ef1335b2c1906abae2f8c5c336891d582bd/modules/masterbar/admin-color-schemes/colors/_admin.scss#L0-L1

2) If the changes affect how new nav is shown in Calypso, we should make sure that the result is what we anticipate by inspecting the changes while nav-unification is enabled (paYJgx-1af-p2)

I will be more than happy to assist you if needed!

sixhours commented:
@cpapazoglou Thank you! I don't think these changes should effect /wp-admin as they're focused on upsells and notices and I understand they don't appear in /wp-admin.

I will make sure to test with nav unification enabled, and I can definitely use your help in testing to make sure I haven't missed something!

sixhours commented:
With nav unification active, I see we need to do some padding adjustments within the nudge:

<img width="289" alt="Screen Shot 2021-01-05 at 10 33 01 AM" src="https://user-images.githubusercontent.com/2124984/103665422-64f24680-4f41-11eb-9393-ef3d67815b58.png">

sixhours committed:
Update domain nudges to use the same style.

sixhours commented:
For the domain nudge, we run out of horizontal space very quickly:

<img width="283" alt="Screen Shot 2021-01-05 at 11 23 53 AM" src="https://user-images.githubusercontent.com/2124984/103671200-97537200-4f48-11eb-91ca-7ebe4caeb38f.png">

This is an issue even now, without the extra outer margins, but it becomes more obvious with them. We could force word wrapping on the domain URL (which is likely to be long) to help this:

<img width="301" alt="Screen Shot 2021-01-05 at 11 20 34 AM" src="https://user-images.githubusercontent.com/2124984/103673366-7ccec800-4f4b-11eb-9e5a-3fb9cd48f1c1.png">

Or maybe use overflow to hide it:

<img width="276" alt="Screen Shot 2021-01-05 at 11 46 56 AM" src="https://user-images.githubusercontent.com/2124984/103673863-1007fd80-4f4c-11eb-812f-d92eebc12698.png">

blackjackkent commented:
I think the overflow approach looks okay. Or maybe have it truncate to x number of characters with a `...` at the end?

ghost commented:
+1 to the truncated version. It's a bit easier to read.

sixhours committed:
Overflow the URL with text-overflow

sixhours commented:
<img width="276" alt="Screen Shot 2021-01-05 at 4 14 10 PM" src="https://user-images.githubusercontent.com/2124984/103699529-1492dd00-4f71-11eb-9f4e-a74684c6143f.png">

This works pretty well!

blackjackkent commented:
Yeah, that looks great I think! :D

sixhours committed:
Minor tweaks to padding/spacing for nav unification variation.

sixhours commented:
@cpapazoglou 👋 I updated the nav unification SCSS in this PR to tweak the padding on the upsell to account for the lack of a border:

Before
<img width="300" src="https://user-images.githubusercontent.com/2124984/103665422-64f24680-4f41-11eb-9393-ef3d67815b58.png">

After
<img width="300" alt="Screen Shot 2021-01-06 at 11 34 47 AM" src="https://user-images.githubusercontent.com/2124984/103795124-976c7400-5013-11eb-8327-c219155b8fe2.png">

Let me know if I've updated the right styles and if I'm missing any test cases to check for visual issues. :)

cpapazoglou commented:
> @cpapazoglou 👋 I updated the nav unification SCSS in this PR to tweak the padding on the upsell to account for the lack of a border:
> 
> Before
> <img alt="" width="300" src="https://user-images.githubusercontent.com/2124984/103665422-64f24680-4f41-11eb-9393-ef3d67815b58.png">
> 
> After
> <img alt="Screen Shot 2021-01-06 at 11 34 47 AM" width="300" src="https://user-images.githubusercontent.com/2124984/103795124-976c7400-5013-11eb-8327-c219155b8fe2.png">
> 
> Let me know if I've updated the right styles and if I'm missing any test cases to check for visual issues. :)

Thanks for wrangling it @sixhours, it looks great and the code is well placed!

kwight commented:
It looks like you're good to go here @sixhours 🎉
